File: legacy/main/rs/emulate/legacy/config/animation/DefaultAnimationDefinition.java
package rs.emulate.legacy.config.animation;

import static rs.emulate.shared.property.Properties.alwaysTrue;
import static rs.emulate.shared.property.Properties.unsignedByte;
import static rs.emulate.shared.property.Properties.unsignedShort;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.function.BiConsumer;
import java.util.function.Function;
import java.util.function.Supplier;

import rs.emulate.legacy.config.ConfigProperty;
import rs.emulate.legacy.config.ConfigPropertyMap;
import rs.emulate.legacy.config.DefaultConfigDefinition;
import rs.emulate.shared.util.DataBuffer;

/**
 * A default {@link AnimationDefinition} used as a base for an actual definitions.
 *
 * @author Major
 */
public class DefaultAnimationDefinition extends DefaultConfigDefinition {

	/**
	 * The DefaultAnimationDefinition.
	 */
	private static final DefaultAnimationDefinition DEFAULT = new DefaultAnimationDefinition();

	/**
	 * A {@link Supplier} that returns a {@link ConfigPropertyMap} copy of this default definition.
	 */
	public static final Supplier<ConfigPropertyMap> SUPPLIER = DEFAULT::toPropertyMap;

	/**
	 * Creates the DefaultAnimationDefinition.
	 */
	private DefaultAnimationDefinition() {
		super();
	}

	@Override
	protected Map<Integer, ConfigProperty<?>> init() {
		Map<Integer, ConfigProperty<?>> properties = new HashMap<>(11);

		properties.put(1, new ConfigProperty<>(AnimationProperty.FRAMES, FrameCollection.EMPTY, FrameCollection::encode,
 FrameCollection::decode, FrameCollection::bytes, input -> Optional.empty())); // XXX

		properties.put(2, unsignedShort(AnimationProperty.LOOP_OFFSET, -1));

		Function<DataBuffer, byte[]> interleaveDecoder = buffer -> {
			int count = buffer.getUnsignedByte();
			byte[] interleaveOrder = new byte[count];

			buffer.get(interleaveOrder);
			return interleaveOrder;
		};

		BiConsumer<DataBuffer, byte[]> interleaveEncoder = (buffer, interleave) -> buffer.putByte(interleave.length).put(
				interleave);

		properties.put(3, new ConfigProperty<>(AnimationProperty.INTERLEAVE_ORDER, new byte[0], interleaveEncoder,
				interleaveDecoder, interleave -> interleave.length + Byte.BYTES, input -> Optional.empty())); // XXX

		properties.put(4, alwaysTrue(AnimationProperty.STRETCHES, false));
		properties.put(5, unsignedByte(AnimationProperty.PRIORITY, 5));
		properties.put(6, unsignedShort(AnimationProperty.PLAYER_MAINHAND, -1));
		properties.put(7, unsignedShort(AnimationProperty.PLAYER_OFFHAND, -1));
		properties.put(8, unsignedByte(AnimationProperty.MAXIMUM_LOOPS, 99));
		properties.put(9, unsignedByte(AnimationProperty.ANIMATING_PRECEDENCE, -1));
		properties.put(10, unsignedByte(AnimationProperty.WALKING_PRECEDENCE, -1));
		properties.put(11, unsignedByte(AnimationProperty.REPLAY_MODE, 2));

		return properties;
	}

}

File: legacy/src/main/rs/emulate/legacy/config/animation/DefaultAnimationDefinition.kt
package rs.emulate.legacy.config.animation

import rs.emulate.legacy.config.DefaultConfigDefinition
import rs.emulate.legacy.config.Properties.alwaysTrue
import rs.emulate.legacy.config.Properties.unsignedByte
import rs.emulate.legacy.config.Properties.unsignedShort
import rs.emulate.legacy.config.SerializableProperty
import rs.emulate.legacy.config.animation.AnimationProperty.*
import rs.emulate.shared.util.DataBuffer
import java.util.HashMap

/**
 * A default [AnimationDefinition] used as a base for an actual definitions.
 *
 * @param T The type of [AnimationDefinition] this default is for.
 */
class DefaultAnimationDefinition<T : AnimationDefinition> : DefaultConfigDefinition<T>() {

    override fun init(): Map<Int, SerializableProperty<*>> {
        val properties = HashMap<Int, SerializableProperty<*>>(11)

        properties[1] = SerializableProperty(FRAMES, FrameCollection.EMPTY, FrameCollection.Companion::encode,
            FrameCollection.Companion::decode, FrameCollection.Companion::bytes
        )

        properties[2] = unsignedShort(LOOP_OFFSET, -1)

        val interleaveDecoder = { buffer: DataBuffer ->
            val count = buffer.getUnsignedByte()
            val interleaveOrder = ByteArray(count)

            buffer.read(interleaveOrder)
            interleaveOrder
        }

        val interleaveEncoder: (DataBuffer, ByteArray) -> DataBuffer = { buffer, order ->
            buffer.putByte(order.size).put(order)
        }

        properties[3] = SerializableProperty(INTERLEAVE_ORDER, ByteArray(0), interleaveEncoder,
            interleaveDecoder, { interleave -> interleave.size + java.lang.Byte.BYTES }
        )

        properties[4] = alwaysTrue(STRETCHES, false)
        properties[5] = unsignedByte(PRIORITY, 5)
        properties[6] = unsignedShort(PLAYER_MAINHAND, -1)
        properties[7] = unsignedShort(PLAYER_OFFHAND, -1)
        properties[8] = unsignedByte(MAXIMUM_LOOPS, 99)
        properties[9] = unsignedByte(ANIMATING_PRECEDENCE, -1)
        properties[10] = unsignedByte(WALKING_PRECEDENCE, -1)
        properties[11] = unsignedByte(REPLAY_MODE, 2)

        return properties
    }

}
